### PR TITLE
[Merged by Bors] - feat: port Data.Sym.Basic

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
There's a dangerous coercion here that I'm not sure how to fix

File: Mathlib/Data/Sym/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+
+/-- Construct an element of the `n`th symmetric power from a multiset of cardinality `n`.
+-/
+@[match_pattern] --Porting note: removed `@[simps]`, generated bad lemma
+abbrev mk (m : Multiset α) (h : Multiset.card m = n) : Sym α n :=
+  ⟨m, h⟩
+#align sym.mk Sym.mk
+
+/-- The unique element in `sym α 0`.
+-/
```
Comment (jcommelin) [inline Mathlib/Data/Sym/Basic.lean:98]:
```suggestion
/-- The unique element in `Sym α 0`. -/
```

File: Mathlib/Data/Sym/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+  rfl
+#align sym.coe_map Sym.coe_map
+
+theorem map_injective {f : α → β} (hf : Injective f) (n : ℕ) :
+    Injective (map f : Sym α n → Sym β n) := fun _ _ h =>
+  coe_injective <| Multiset.map_injective hf <| coe_inj.2 h
+#align sym.map_injective Sym.map_injective
+
+/-- Mapping an equivalence `α ≃ β` using `sym.map` gives an equivalence between `sym α n` and
+`sym β n`. -/
```
Comment (jcommelin) [inline Mathlib/Data/Sym/Basic.lean:413]:
```suggestion
/-- Mapping an equivalence `α ≃ β` using `Sym.map` gives an equivalence between `Sym α n` and
`Sym β n`. -/
```

File: Mathlib/Data/Sym/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+#align sym.attach_nil Sym.attach_nil
+
+@[simp]
+theorem attach_cons (x : α) (s : Sym α n) :
+    (cons x s).attach =
+      cons ⟨x, mem_cons_self _ _⟩ (s.attach.map fun x => ⟨x, mem_cons_of_mem x.prop⟩) :=
+  coe_injective <| Multiset.attach_cons _ _
+#align sym.attach_cons Sym.attach_cons
+
+/-- Change the length of a `sym` using an equality.
```
Comment (jcommelin) [inline Mathlib/Data/Sym/Basic.lean:462]:
```suggestion
/-- Change the length of a `Sym` using an equality.
```

File: Mathlib/Data/Sym/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+
+theorem mem_fill_iff {a b : α} {i : Fin (n + 1)} {s : Sym α (n - i)} :
+    a ∈ Sym.fill b i s ↔ (i : ℕ) ≠ 0 ∧ a = b ∨ a ∈ s := by
+  rw [fill, mem_cast, mem_append_iff, or_comm, mem_replicate]
+#align sym.mem_fill_iff Sym.mem_fill_iff
+
+open Multiset
+
+/-- Remove every `a` from a given `sym α n`.
+Yields the number of copies `i` and a term of `sym α (n - i)`. -/
```
Comment (jcommelin) [inline Mathlib/Data/Sym/Basic.lean:543]:
```suggestion
/-- Remove every `a` from a given `Sym α n`.
Yields the number of copies `i` and a term of `Sym α (n - i)`. -/
```

**Review by jcommelin — COMMENTED** [review]
I've reviewed for naming issues in comments.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/Data/Sym/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+#align sym.sym' Sym.Sym'
+
+/-- This is `cons` but for the alternative `Sym'` definition.
+-/
+def cons' {α : Type _} {n : ℕ} : α → Sym' α n → Sym' α (Nat.succ n) := fun a =>
+  Quotient.map (Vector.cons a) fun ⟨_, _⟩ ⟨_, _⟩ h => List.Perm.cons _ h
+#align sym.cons' Sym.cons'
+
+@[inherit_doc]
+notation a "::" b => cons' a b
```
Comment (gebner) [inline Mathlib/Data/Sym/Basic.lean:246]: Overloading `::` broke pattern-matching btw. #1735

**Pull request closed without merging**